lib/streamlit/hello.py

```diff
@@ -17,55 +17,172 @@
+    """
+    red = st.sidebar.slider("Red", 0, 255, 0)
+    green = st.sidebar.slider("Green", 0, 255, 128)
+    blue = st.sidebar.slider("Blue", 0, 255, 0)
+    array = np.zeros([512, 512, 3], dtype=np.uint8)
+    array[:, :, 0].fill(red)
+    array[:, :, 1].fill(green)
+    array[:, :, 2].fill(blue)
+    st.markdown("#### RGB color (%d, %d, %d)" % (red, green, blue))
+    st.image(array)
 
-        # Update status text.
-        status_text.text("The latest random number is: %s" % new_rows[-1, 0])
 
-        # Append data to the chart.
-        chart.add_rows(new_rows)
+# Data for demo_5 derived from
+# https://www.kaggle.com/rounakbanik/the-movies-dataset
+DATASET_URL = (
+    "https://streamlit-demo-data.s3-us-west-2.amazonaws.com/movies-revenue.csv.gz"
+)
 
-        # Pretend we're doing some computation that takes time.
-        time.sleep(0.1)
 
-    status_text.text("Done!")
-    st.balloons()
+def demo_5():
+    """
+    Discover the gross revenue of a movie of your liking!
+    Note that the network loading and preprocessing of the data is cached
+    by using Streamlit st.cache annotation.
+    """
+    import pandas as pd
+
+    @st.cache
+    def load_dataframe_from_url():
+        return pd.read_csv(DATASET_URL).transform(
+            {"title": lambda x: x, "revenue": lambda x: round(x / 1000 / 1000)}
+        )
+
+    df = load_dataframe_from_url()
+    movie = st.text_input("Search movie titles")
+    df = df[df.title.str.contains(movie, case=False)].sort_values(
+        ascending=False, by="revenue"
+    )
+    st.markdown("#### Result dataframe for %d movie(s)" % df.shape[0])
+    st.dataframe(df.rename(columns={"title": "Title", "revenue": "Revenue [$1M]"}))
+
 
+DEMOS = OrderedDict(
+    {
+        "---": intro,
+        "Basic Text": demo_1,
+        "Simple Interaction": demo_2,
+        "Animation": demo_3,
+        "Sidebar": demo_4,
+        "Caching": demo_5,
+    }
+)
+
+
+def run():
+    st.title("Welcome to Streamlit!")
     st.write(
         """
-        PS: Want to know how we did this?
-        [You can learn about it here.](https://streamlit.io/secret/docs/core_mechanics.html)
-    """
+        Streamlit is the best way to create bespoke apps.
+        """
     )
 
+    demo_name = st.selectbox("Choose a demo", list(DEMOS.keys()), 0)
+    demo = DEMOS[demo_name]
+
+    if demo_name != "---":
+        st.markdown("## %s Demo" % demo_name)
+        st.write(inspect.getdoc(demo))
+        st.write("### Code")
+        sourcelines, n_lines = inspect.getsourcelines(demo)
+        sourcelines = reset_indentation(remove_docstring(sourcelines))
+        st.code("".join(sourcelines))
+        if st.checkbox("Run %s Demo" % demo_name):
+            st.write("---\n### Execution")
+            demo()
+    else:
+        demo()
+
+
+# This function parses the lines of the function and removes the docstring
+# if found.
+def remove_docstring(lines):
+    if len(lines) < 3 and '"""' not in lines[1]:
```

Comment:
This case is for when there is no docstring? But we also expect the number of lines to be less than 3? 

Couldn't we have no docstring with more lines? Or are we expecting not to have to handle this case?

Reply:
Thanks... this is really helpful. This is a bit hacky and I need to refine it to work in the general case.

I am bailing removing the docstring if a specific pattern cannt be detected. In this case,
the reasoning behind this is i) if lines[1] is not a `"""`, then there is no docstring. I also want another line for a closing docstring, so that I can start parsing from line 2, so if `len(lines) < 3` just bail. 

If `lines[2]` is a `"""`, the `while` loop will never be entered and `index` is going to be `2`, so I return lines[2 + 1 :] which probably breaks (may need to fix this)